internal/multirun.py

```diff
@@ -29,46 +35,50 @@ def _run_command(command: Command, block: bool, **kwargs) -> Union[int, subproce
         args = [command.path] + command.args
     env = dict(os.environ)
     env.update(command.env)
-    if block:
+    if check_call:
         return subprocess.check_call(args, env=env)
     else:
-        return subprocess.Popen(args, env=env, **kwargs)
+        p = subprocess.Popen(args, env=env, **kwargs)
+        started_processes.append(p)
+        stdout, stderr = p.communicate()
+        return p, stdout, stderr
 
 
-def _perform_concurrently(commands: List[Command], print_command: bool, buffer_output: bool) -> bool:
+def _perform_concurrently(commands: List[Command], concurrency: int, print_command: bool, buffer_output: bool) -> bool:
     kwargs = {}
     if buffer_output:
         kwargs = {
              "stdout" : subprocess.PIPE,
              "stderr" : subprocess.STDOUT
         }
 
-    processes = [
-        (command, _run_command(command, block=False, **kwargs))
-        for command
-        in commands
-    ]
+    with ThreadPool(concurrency) as tp:
+        started_processes = []
+        results = [
+            (command, tp.apply_async(_run_command, args=(command, False, started_processes), kwds=kwargs))
+            for command
+            in commands
+        ]
 
-    success = True
-    try:
-        for command, process in processes:
-            process.wait()
```

Comment:
As described by Python doc:

> Note This will deadlock when using stdout=PIPE or stderr=PIPE and the child process generates enough output to a pipe such that it blocks waiting for the OS pipe buffer to accept more data. Use [Popen.communicate()](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate) when using pipes to avoid that.